pkg/capabilities/consensus/ocr3/requests/handler.go

```diff
@@ -0,0 +1,146 @@
+package requests
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+type responseCacheEntry struct {
+	response  *Response
+	entryTime time.Time
+}
+
+type Handler struct {
+	services.StateMachine
+
+	lggr logger.Logger
+
+	mux             sync.Mutex
+	store           *Store
+	responseCache   map[string]*responseCacheEntry
+	cacheExpiryTime time.Duration
+
+	stopCh services.StopChan
+	wg     sync.WaitGroup
+}
+
+func NewHandler(lggr logger.Logger, s *Store, responseExpiryTime time.Duration) *Handler {
+	return &Handler{
+		lggr:            lggr,
+		store:           s,
+		responseCache:   map[string]*responseCacheEntry{},
+		cacheExpiryTime: responseExpiryTime,
+		stopCh:          make(services.StopChan),
+	}
+}
+
+func (h *Handler) SendResponse(resp *Response) {
+	req := h.getRequestForResponse(resp)
+	if req != nil {
+		req.sendResponse(resp)
+	}
+}
+
+func (h *Handler) SendRequest(r *Request) error {
+	resp, err := h.addRequest(r)
+	if err != nil {
+		return fmt.Errorf("failed to add request: %w", err)
+	}
+
+	if resp != nil {
+		r.sendResponse(resp)
+		return nil
+	}
+
+	return nil
+}
+
+func (h *Handler) ResponseCacheSize() int {
+	h.mux.Lock()
+	defer h.mux.Unlock()
+
+	return len(h.responseCache)
+}
+
+func (h *Handler) Start(_ context.Context) error {
+	return h.StartOnce("RequestHandler", func() error {
+		h.wg.Add(1)
+		go func() {
+			defer h.wg.Done()
+			ctx, cancel := h.stopCh.NewCtx()
+			defer cancel()
+			h.checkForResponseExpiry(ctx)
+		}()
+		return nil
+	})
+}
+
+func (h *Handler) Close() error {
+	return h.StopOnce("RequestHandler", func() error {
+		close(h.stopCh)
+		h.wg.Wait()
+		return nil
+	})
+}
+
+func (h *Handler) checkForResponseExpiry(ctx context.Context) {
+	ticker := time.NewTicker(h.cacheExpiryTime)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			h.expireCachedResponses()
+		}
+	}
+}
+
+func (h *Handler) expireCachedResponses() {
+	h.mux.Lock()
+	defer h.mux.Unlock()
+
+	for k, v := range h.responseCache {
+		if time.Since(v.entryTime) > h.cacheExpiryTime {
+			delete(h.responseCache, k)
+			h.lggr.Debugw("Expired response", "workflowExecutionID", k)
+		}
+	}
+}
+
+func (h *Handler) getRequestForResponse(resp *Response) *Request {
```

Comment:
Name of this function should reflect what it does. It's not only a getter, it actually does all the magic - caches the response for the future :)